src/AppMenuSearch.cc

```
@@ -263,9 +263,31 @@ void AppMenuSearch::collectSearchCandidates(QMenu *menu, QSet<QMenu *> &visited,
     }
 }
 
-bool AppMenuSearch::matchesAncestorsOrText(const SearchCandidate &candidate, const QString &itemText, const QStringMatcher &matcher, bool ignoreTopLevel, QHash<QAction *, MatchState> &matchCache) const
+bool AppMenuSearch::matchesAncestorsOrText(const SearchCandidate &candidate, const QString &itemText, const QStringMatcher &matcher, bool ignoreTopLevel, QHash<QAction *, MatchState> &matchCache, QHash<QAction *, bool> &pathMatchCache) const
 {
+    // 1. O(1) Fast-Path: check if the direct parent menu's path evaluation is already cached.
+    // Safe because each submenu action occurs in one unique root-to-leaf path (since Qt's QObject/QMenu
+    // hierarchy forms a strict, tree-structured parent-child relationship with unique parentage).
+    QAction *lastAncestor = candidate.ancestors.isEmpty() ? nullptr : candidate.ancestors.last();
```

Clang (14.0.6)

[warning] variable 'lastAncestor' is not initialized

(cppcoreguidelines-init-variables)